feat(Tactic/Linter): lint unwanted unicode

adomasbaliuka opened the pull request:
Add a text-based style linter that checks all unicode characters. 

- Only allow whitelisted characters, with a suggestion to add the flagged character to the whitelist
- Some characters are specified to use emoji-variant or text-variant. The linter ensures these variant-selectors only appear in the correct places and provides an auto-fix to clean them up.

---

Discussed at [Zulip](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Whitelist.20for.20Unicode.3F/near/448721083)

The current proposal checks each character against a whitelist. There is also the question of what the whitelist should contain. 

A reasonable whitelist might be "all unicode currently in Mathlib" ∪ [Wikipedia/Mathematical_operators_and_symbols_in_Unicode](https://en.wikipedia.org/wiki/Mathematical_operators_and_symbols_in_Unicode) ∪ [EdAyers/mathlib3/docs/unicode.md](https://github.com/EdAyers/mathlib/blob/fb5c2ffe917049959597b0a93b872e91f4004922/docs/unicode.md). Currently implemented is a subset of this list closer to what's currently present in mathlib.

- [x] depends on: #17068
- [x] depends on: #17116
- [x] depends on: #17129 (part 1 of this PR)
- ~depends on #34719 (part 1 of this PR)~
-   ~depends on: #17131 (part 2 of this PR)~ (obsolete)
- [x]  depends on: #34022 (part 2 of this PR)
- Future note: @adomasbaliuka says: "we might add something to the Mathlib 4 docs about this as well?"

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

adomasbaliuka committed:
Adds first idea for text-based linter that flags unicode outside a whitelist

adomasbaliuka committed:
Adds comments to definitions

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -103,6 +107,8 @@ def StyleError.errorMessage (err : StyleError) (style : ErrorFormat) : String :=
         s!"{sizeLimit} file contains {currentSize} lines, try to split it up"
     | ErrorFormat.humanReadable => s!"file contains {currentSize} lines, try to split it up"
+  | StyleError.newUnicode c =>
+      s!"Unicode character {c} is not recommended. Consider adding it to the whitelist."
```
```suggestion
      s!"unicode character '{c}' is not recommended. Consider adding it to the whitelist."

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -114,6 +120,7 @@ def StyleError.errorCode (err : StyleError) : String := match err with
   | StyleError.broadImport _ => "ERR_IMP"
   | StyleError.lineLength _ => "ERR_LIN"
   | StyleError.fileTooLong _ _ _ => "ERR_NUM_LIN"
+  | StyleError.newUnicode _ => "NO_ERR_CODE_AVAILABLE"
```
You can just make one up, perhaps "ERR_UNICODE".

Can you also update the logic for `parse?_errorContext` below, please? (You can follow how e.g. the `ERR_AUT` code is treated.)

adomasbaliuka replied:
I put the "test" I made for this `parse?_errorContext` into ` test/LintStyle.lean`.

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -380,9 +387,44 @@ def checkFileLength (lines : Array String) (existingLimit : Option ℕ) : Option
+
+section unicodeLinter
+
+/- Hashable instance for use with the unicodeLinter -/
```
Does this fix it? A /- Comment -/ is just a multi-line comment in Lean, a /-- Comment -/ is a documentation comment.
```suggestion
/- Hashable instance for use with the unicodeLinter -/

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -380,9 +387,44 @@ def checkFileLength (lines : Array String) (existingLimit : Option ℕ) : Option
+local instance : Hashable Char where
+  hash c := c.val.toUInt64
+
+/-- TODO make complete and order nicely -/
```
Yes, ordering this would be great. I presume this contains numbers, all ASCII letters, space etc., but it's very hard for me to check (short of using the search function ~60 times).

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -380,9 +387,44 @@ def checkFileLength (lines : Array String) (existingLimit : Option ℕ) : Option
+    let badChars := line.toList.filter (fun c => !unicodeWhitelist.contains c)
+    if badChars.length > 0 then
+      let newErrors : Array (StyleError × ℕ) :=
+        badChars.toArray.map fun c ↦ (StyleError.newUnicode c, lineNumber)
```
Here's a small Lean trick, just so you know.
You can also use \centerdot to write this more compactly as `badChars.toArray.map (StyleError.newUnicode ·, lineNumber)`; same for the definition of `badChars` three lines higher.

grunweg commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -6,6 +6,7 @@ Authors: Michael Rothgang
 
 import Batteries.Data.String.Matcher
 import Mathlib.Data.Nat.Notation
+import Batteries.Lean.HashSet
```
Can you sort the imports alphabetically, if possible? Thanks!

grunweg reviewed (commented):
Some quick comments. Overall, this linter
- flags interesting things
- does not seem to turn up something worrying,
both of which is great.

grunweg commented:
Would you like to take a look at the "spaces" the linter flags? (I haven't checked what this is about exactly; just looking at it in the text editor is probably not useful...)

adomasbaliuka committed:
Update linter warning message

Co-authored-by: grunweg <rothgami@math.hu-berlin.de>

adomasbaliuka committed:
Implements some reviewer comments

adomasbaliuka committed:
Updates whitelist

adomasbaliuka commented:
> Would you like to take a look at the "spaces" the linter flags? (I haven't checked what this is about exactly; just looking at it in the text editor is probably not useful...)

That seems to be

 ' ': Unicode U+202F (category Zs: Separator, space)

I think we'll remove them from Mathlib but let's see...

adomasbaliuka committed:
Adds missing docstring

adomasbaliuka committed:
Splits up sections of whitelist instead of doing in-line comments in the string

adomasbaliuka committed:
Adds unicode value (in hex) as part of linter warning

adomasbaliuka committed:
Renames and fixes function printCodepointHex

adomasbaliuka commented:
I have removed all non-breaking spaces (U+00a0) from Mathlib. This might be excessive:
- I can PR it separately if the change is too big or off topic in this PR
- We may decide to *allow* non-breaking spaces

grunweg commented:
Thank you! I think splitting the removal of non-breaking spaces into a separate PR (do you know git cherry-pick?) makes sense, and would be happy to nominate it for the maintainers to look. Reading the zulip discussion, changing the spaces does not appear controversial to me.

adomasbaliuka committed:
Fix $ vs <| preference as some linter wants

adomasbaliuka committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter

adomasbaliuka committed:
Moves test for parse?_errorContext to test/LintStyle.lean

adomasbaliuka committed:
merge master

adomasbaliuka committed:
Removes U+200C (Zero Width Non-Joiner ZWNJ) from unicode whitelist

Also fix some comments

adomasbaliuka committed:
Removes U+FE0F (Variation Selector-16 VS16) from unicode whitelist

adomasbaliuka committed:
Makes whitelists sections into Array Char. Removes some more weird characters

adomasbaliuka committed:
Fixes comment

adomasbaliuka commented:
I split up the whitelist strings into `Array Char` because this way it's more clear what the individual characters are.

I also removed some more weird characters (invisible, or ones that my VSCode font does not support). We can add them back if needed once most of the non-breaking whitespace is gone...

adomasbaliuka committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter

joneugster committed:
add logic for unicode variant selectors

joneugster commented:
@adomasbaliuka I've added the logic to validate the variant-selector characters here: #16368

Maybe you could have a look, especially at the changes in [`TextBased.lean`](https://github.com/leanprover-community/mathlib4/pull/16368/files#diff-c973e59ceee698f73d927eed559746d5b761304155603aac9f1f41f7375b7e85) and if you think they are reasonable, I could directly push them to this PR.

In particular it showed me that there is already one place in mathlib where there are characters followed by both selectors `\uFE0E\FE0F`, which is not valid Unicode, so having a linter for this would certainly be good.

joneugster committed:
fixes

adomasbaliuka committed:
Fixes docstring

joneugster commented on Mathlib/Init/Data/String.lean:
I didn't know where to put these, so I put them here for now.

I'll make a PR to Lean core to modify `String.get` now, so hopefully `get₂` isn't required by the time this is up for review.

joneugster replied:
(we dont actually need this to be changed in core. Replaced the one remaining occurence with `s.get? _ |>.getD '\uFFFD')`

joneugster commented:
@adomasbaliuka I've pushed my suggestions from #16368 here, feel free to take it from here and make changes as you see fit!

I left the lint errors unfixed so you have a chance to look at them, I figured it's only worth fixing them once your happy with the linter

adomasbaliuka commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -371,9 +431,231 @@ def checkFileLength (lines : Array String) (existingLimit : Option ℕ) : Option
+      for pos in charPos do
+        let nextC := line.get₂ (line.next pos)
+        let previousC := line.get₂ (line.prev pos) -- may be the char. itself if `pos == 0`
+        unless (previousC == '\'' ∧ nextC == '\'') ∨ nextC == selector do
```
What is the logic here? What do the ' characters signify?

joneugster replied:
outdated now

adomasbaliuka committed:
Completely reworks unicode linter

Using whitelist to ignore characters, using further lists to check emoji/text variant selector use

adomasbaliuka commented:
@joneugster  I completely changed the way it's implemented (but I think it should do the same thing as your version, except about the ' characters where I left a comment). 

The whitelist is now only used to ignore characters. Special cases such as variant selectors and emojis are checked further and then allowed despite not being in the whitelist if they are used correctly. This way we have only one lookup in the whitelist for each whitelisted character, which should be faster (not sure how relevant speed is but I hope this code is clearer).

There may be an issue with the usage of the `next` and `prev` functions. They operate on UTF8 bytes and give unspecified results if given an invalid position (which I'm not sure can happen with the current code but I think yes?!).

I'm also not sure about which characters exactly should have which selectors if any. Some arrows in Mathlib have text selectors, for example. For several characters in the emoji-list, the selectors don't seem to be doing anything. Can we check this once more?

adomasbaliuka committed:
Fixes issue where endPos was given to String.next

adomasbaliuka committed:
Removes test-like thing that was checking effect of selectors.

The effect depends on the font that is used.

adomasbaliuka committed:
Adds test. Improves error message (missing selector vs wrong selector)

joneugster committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter

adomasbaliuka commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +377,507 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+  return errors
+
+
+namespace unicodeLinter
```
This stuff is duplicated here

joneugster replied:
I believe this is fixed

joneugster committed:
updating iteration logic

joneugster committed:
run mk_all

joneugster committed:
fix imports

joneugster commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +379,261 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+    (err : Array StyleError := #[]) : Array StyleError :=
+  if h : pos < s.endPos then
+    have := Nat.sub_lt_sub_left h (String.lt_next s pos)
+    let posₙ := s.next pos
+    let cₙ := s.get? posₙ |>.getD '\uFFFD' -- �
```
This is the only occurence of `String.next`. @adomasbaliuka I guess this is fine, isn't it?

adomasbaliuka replied:
Seems fine, assuming the function is never called with an invalid position. Maybe put that in the docstring? (we inherit the fact that the result is "unspecified" for invalid `pos`)

adomasbaliuka replied:
I did it.

joneugster committed:
cleanup and fixes

joneugster committed:
remove unused import

joneugster committed:
update test

joneugster commented:
@adomasbaliuka What do you think of the version I now pushed? Maybe you can test for speed again. This way should be even less looping and it's overall a bit shorter.

Also, maybe you know better how to write the `parse?_errorContext` entry?

Let me know what you think, and if you're happy I could clean up the text and tests and fix the linter warnings

joneugster commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +378,260 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+'\u23F3',        -- ⏳️
+.ofNat 0x1F3C1 ] -- 🏁️
+
+/-- Unicode symbols in mathilb that should always be followed by the text-variant selector. -/
+def nonEmojis : Array Char := #[]
+-- TODO: should there be any? maybe #['↗', '↘', '✝', '▼', '▶']?
+
+/--
+Other unicode characters present in Mathlib (as of Aug. 28, 2024)
+not present in any of the above lists.
+-/
+def othersInMathlib := #[
+'▼', 'c', 'ō', '⏩', '❓',
+'🆕', 'š', 'ř', '⚬', '│', '├', '┌', 'ő',
+'⟍', '̂', 'ᘁ', 'ń', 'ć', '⟋', 'ỳ', 'ầ', '⥥',
+'ł', '◿', '◹', '－', '＼', '◥', '／', '◢', 'Ž', 'ă',
+'И', 'в', 'а', 'н', 'о', 'и', 'ч', 'Š', 'ᴜ', 'ᵧ', '´',
+'ᴄ', 'ꜰ', 'ß', 'ᴢ', 'ᴏ', 'ᴀ', 'ꜱ', 'ɴ', 'ꞯ', 'ʟ',
+'ʜ', 'ᵟ', 'ʙ', 'ᵪ', 'ᵩ', 'ᵦ', 'ᴊ', 'ᴛ', 'ᴡ', 'ᴠ',
+'ɪ', '̀', 'ᴇ', 'ᴍ', 'ʀ', 'ᴅ', 'ɢ', 'ʏ', 'ᴘ', 'ĝ', 'ᵨ',
+'ᴋ', 'ś', '꙳', '𝓡', '𝕝', '𝖣', '⨳' ]
```
I think it's fine to remain conservative with the non-emoji characters, and even leave it empty for now.

And regarding your question further up, I think it's fine if you add whatever whitelist you like best to the `othersInMathlib`. Although, it's also desirable that this list here eventually decreases in size in favour of adding VSCode abbreviations for the characters.

adomasbaliuka replied:
Leaving the non-emojis empty is fine for me.

I don't understand the other point, sorry. Add what whitelist to what?

joneugster committed:
fix ASCII range

joneugster commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +379,261 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+def ASCII.whitespace (c : Char) := #[' ', '\n'].contains c
+
+/-- Printable ASCII characters, not including whitespace. -/
+def ASCII.printable (c : Char) := 21 ≤ c.toNat || c.toNat ≤ 126
```
ofc thanks for spotting🤦🏻‍♂️🤦🏻‍♂️

adomasbaliuka commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +378,260 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+  hash c := c.val.toUInt64
+
+/-- Allowed whitespace characters -/
+def ASCII.whitespace (c : Char) := #[' ', '\n'].contains c
```
I would call this e.g. `allowedWhitespace` so as to avoid confusion with the concept "all ASCII whitespace".

adomasbaliuka replied:
I renamed it.

adomasbaliuka commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -190,6 +246,15 @@ def parse?_errorContext (line : String) : Option ErrorContext := Id.run do
+            else none
+        | "ERR_UNICODE_VARIANT" =>
+          -- some (StyleError.unicodeVariant _ _ _)
+          none --TODO
```
I think the idea here is to parse back the text representation of the linter error into an `ErrorContext` object. I don't actually know what it's used for. If I understand it correctly, it should be the inverse of `outputMessage`, which actually isn't strictly possible the way that we print the error (we ignore the second character, for example). 

Probably this doesn't matter and we just "best-effort" parse the error message.  I could do it if you like...

adomasbaliuka replied:
Did this.

adomasbaliuka commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -314,9 +378,260 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+      -- okay
+      findBadUnicodeAux s posₙ cₙ err
+  -- emojis/non-emojis should not be the last character in the line
+  -- in practise, the last char is `\n`, so this is in theory superfluous.
```
I believe with the way this linter is called, the final \n is not part of the "line" we process.

adomasbaliuka replied:
I removed the comment and kept the check. Please double-check if you agree.

adomasbaliuka commented on test/LintStyle.lean:
```diff
@@ -135,3 +135,18 @@ lemma foo' : True := trivial
+-- The list `othersInMathlib` should only contain characters not present in another list.
+#guard unicodeLinter.othersInMathlib.toList ∩ unicodeLinter.withVSCodeAbbrev.toList = ∅
+#guard unicodeLinter.othersInMathlib.toList ∩ unicodeLinter.emojis.toList = ∅
+#guard unicodeLinter.othersInMathlib.toList ∩ unicodeLinter.nonEmojis.toList = ∅
```
Maybe we should add some more here? Intersection of `emojis` and `nonEmojis` for example?

adomasbaliuka replied:
I added some.

adomasbaliuka committed:
Minor changes

Renames ASCII.whitespace to ASCII.allowedWhitespace.
Adds comments.
Makes sure (by inspection) that results using String positions are never unspecified.
Removes unused Hashable instance for Char.

adomasbaliuka committed:
Removes comment about linefeed being last character

We never see those characters with the way the linter is called

adomasbaliuka committed:
Adds test that special char lists are disjoint

adomasbaliuka committed:
Remove stray character c from List.
Add definition for ASCII.allowed

adomasbaliuka committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter

adomasbaliuka committed:
Implements parsing back error messages for unicodeVariant

I still don't really know what this parsing back is useful for.
Added tests to show it works.
Tested also on current output of linter, when modifying error messages to be in `.exceptionsFile` format.

joneugster committed:
cleanup

joneugster committed:
cleanup

joneugster committed:
add logic for autofix

joneugster committed:
clean up auto-fix

joneugster committed:
fix

joneugster committed:
apply suggestions from 'lake exe lint-style --fix'

joneugster committed:
fix bad whitespace

joneugster committed:
Merge branch 'eugster/bad-whitespace-2409' into adomas_unicodeLinter

joneugster committed:
revert erronous fixes

joneugster committed:
fix

joneugster committed:
apply suggestions from 'lake exe lint-style --fix'

adomasbaliuka commented:
@joneugster LGTM

joneugster committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter

eric-wieser commented on Mathlib/GroupTheory/GroupExtension/Defs.lean:
```diff
@@ -22,14 +22,14 @@ such as splittings and equivalences.
-      ↗︎ E  ↘
+      ↗ E  ↘
 1 → N   ↓    G → 1
-      ↘︎ E' ↗︎️
+      ↘ E' ↗
 
 For additive groups:
-      ↗︎ E  ↘
+      ↗ E  ↘
 0 → N   ↓    G → 0
-      ↘︎ E' ↗︎️
```
I think these are genuinely intended to be non-breaking; the right solution is to put this in a code block.

joneugster replied:
Sorry these are about the text-variant-selector of the arrows characters, i.e. there were invisible `\uFE0E` following the diagonal arrows to prevent them form being displayed as emojis.

There are no spaces changed here.

If they are supposed to have the variant-selector (which is a fair choice) then, they can be added to the list of text-symbols defined here and the selectors will be added back automatically by `lake exe lint-style --fix`

eric-wieser commented on Mathlib/Tactic/Widget/Calc.lean:
```diff
@@ -116,7 +116,7 @@ def suggestSteps (pos : Array Lean.SubExpr.GoalsLocation) (goalType : Expr) (par
 def CalcPanel.rpc := mkSelectionPanelRPC suggestSteps
   "Please select subterms."
-  "Calc 🔍"
+  "Calc 🔍️"
```
These are also intended to be non-breaking; I think using the unicode escape would be appropriate here.

joneugster replied:
same here, there is no space change, but the magnifying glass received a emoji-variant-selector because its specified in the list of emojis.

joneugster replied:
For context, there are currently no non-breaking spaces in any parts of mathlib except in docstrings about condensed mathematics & cat. theory because Dagur's IDE automatically adds them after each closing \`

And the dependency PRs fixing whitespace are mainly because Dagur's quite busy and commiting a lot :D

eric-wieser commented on Mathlib/Tactic/Linter/TextBased.lean:
````diff
@@ -349,12 +427,290 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
+Taken from abbreviations.json in
+github.com/leanprover/vscode-lean4/blob/97d7d8c382d1549c18b66cd99ab4df0b6634c8f1
+
+Obtained using Julia code:
+```julia
+filter(!isascii, unique( <all text in JSON file> )) |> repr
+```
````
Let's do this with Lean!

eric-wieser commented on Mathlib/Tactic/Linter/TextBased.lean:
```diff
@@ -349,12 +427,290 @@ def isImportsOnlyFile (lines : Array String) : Bool :=
 end
 
+
+namespace unicodeLinter
```
This is really quite long; can it go in its own file?

joneugster replied:
absolutely! I'll move it to `Mathlib/Tactic/Linter/Unicode.lean`

joneugster committed:
auto-fix for non-breaking-spaces

joneugster committed:
fixes by 'lake exe lint-style --fix'

joneugster committed:
move unicode linter declarations to their own file

joneugster committed:
Merge remote-tracking branch 'origin/master' into adomas_unicodeLinter
